pkg/reconciler.v1/common/job.go

```diff
@@ -0,0 +1,478 @@
+// Copyright 2021 The Kubeflow Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package common
+
+import (
+	"context"
+	"fmt"
+	"reflect"
+	"strings"
+	"time"
+
+	commonv1 "github.com/kubeflow/common/pkg/apis/common/v1"
+	"github.com/kubeflow/common/pkg/core"
+	commonutil "github.com/kubeflow/common/pkg/util"
+	"github.com/kubeflow/common/pkg/util/k8sutil"
+
+	"github.com/sirupsen/logrus"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+const (
+	GroupName = "kubeflow.org"
+
+	ReasonKey        = "reason"
+	ReasonJobDeleted = "job deleted"
+
+	MsgReconcileCancelled = "Reconcile Cancelled"
+	MsgReconcileStart     = "Reconcile Starts"
+
+	MsgGetPodsFailed     = "Get Pods Failed"
+	MsgGetServicesFailed = "Get Services Failed"
+
+	MsgBackoffLimitReachedTemplate   = "Job %s has failed because it has reached the specified backoff limit"
+	MsgActiveDeadlineReachedTemplate = "Job %s has failed because it was active longer than specified deadline"
+
+	ErrUpdateJobConditionsFailed = "failed to update job conditions"
+
+	ErrUpdateJobErrorTemplate                    = "UpdateJobStatus error %v"
+	ErrAppendJobConditionTemplate                = "Append job condition error %v"
+	ErrReconcilePodsTemplate                     = "ReconcilePods error %v"
+	ErrReconcileServicesTemplate                 = "ReconcileServices error %v"
+	ErrReconcileGangTemplate                     = "ReconcilePodGroups error %v"
+	ErrGetReplicasStatusFromStatusFailedTemplate = "failed to get ReplicasStatus for %s from status"
+
+	WarnDefaultImplementationTemplate = "Warning: executing default implementation for KubeflowReconciler.%s"
+	WarnNotCountedInBackoffLimit      = "The restart policy of replica %v of the job %v is not OnFailure or Always. Not counted in backoff limit."
+)
+
+// KubeflowJobReconciler defines a Reconciler dealing with KubeflowJob
+type KubeflowJobReconciler struct {
+	client.Client
+	ReconcilerUtilInterface
+	PodInterface
+	ServiceInterface
+	GangSchedulingInterface
+	counter *commonutil.Counter
+}
+
+// BareKubeflowJobReconciler returns the pointer of a KubeflowJobReconciler with minimal implementation
+func BareKubeflowJobReconciler(client client.Client) *KubeflowJobReconciler {
```

Comment:
Should we use term like "Base"

Reply:
While the `Bare` means only interface independently defined in `JobInterface` is implemented in the returned `KubeflowJobReconciler` pointer, I think the `Base` conveys the same message.